### ISSUE TITLE
How to manipulate the z vector in Image2Image.py

### ISSUE THREAD

**Issue by kaihuchen** [ppwwyyxx/tensorpack#85]
@ppwwyyxx  I would like to get at the value of the z vector (i.e., the output  of the **e8** node) in Image2Image.py, what would be the best way for me to do this?

The context of this request is this: I already have the cGAN generator trained with satisfactory result. Now I want to run some test samples through this generator, and as each test sample is processed by the generator, I want to get and store the value of the corresponding z vector in a file for further processing. The reason for wanting to do this is that this z vector could be a good representation for the test input image in the latent space, and having it will allow me to do all kinds of things

A second related question is that I also want to inject my own made-up z vector into such a trained generator (bypassing the **conv** part, i.e., the nodes e1-e8 in the generator) in order to create an image out of it. What would be the best way to do this?

**Comment by ppwwyyxx** [comment]
1. In `sample()`, just add `conv8/output` to `output_names`. Then `o[0][:, :, :, ::-1]` would contain 6 BGR images, and `o[1]` would be 6 e8 tensors (of shape 6x512x1x1).

2. You can look at `DCGAN-CelebA.py` which uses an input z vector.

**Comment by ppwwyyxx** [comment]
Sorry I misunderstood your question. I thought you want to use a random z vector.
You can inject your own e8 vector, but the generator still will need e7 ~ e1. I expect changing e8 won't make much difference.

**Comment by kaihuchen** [comment]
@ppwwyyxx 
**On the first question**, I got it to work as you suggested, with the twist that I appended **gen/conv8/output** to output_names to get a 6x1x1x512 tensor, which does appear to be what I wanted. Much appreciated for your help!

One curious thing is that when I visualize the result of >1000 z vectors using t-SNE, I thought the diagram looks different (aside from coloring) from what I got with exactly the same training/testing datasets (which are frames from a celebrity interview video) using pix2pix/Torch.  Not sure what to make of it yet. Something interesting for me to look into.

Following is the tensorpack version of the t-SNE diagram, where the test samples seem fairly uniformly distributed in the z space (projected from 512D to 3D for display):
![tpack](https://cloud.githubusercontent.com/assets/13685803/21621474/2d2e0b9a-d1c7-11e6-888e-921a20cfff39.gif)
 
Following is the pix2pix/Torch version of the t-SNE diagram, which looks comparatively more lumpy:
![ex2](https://cloud.githubusercontent.com/assets/13685803/21621711/2e713fbc-d1c8-11e6-81ac-e65bc10c5aa7.gif)
